types/src/token/account.rs

```diff
@@ -0,0 +1,169 @@
+#[archive_attr(derive(CheckBytes))]
+pub enum Account {
+    /// An externally owned account.
+    External(PublicKey),
+    /// A contract account.
+    Contract(ContractId),
+}
+
+impl Account {
+    const SIZE: usize = 194;
+
+    /// Convert the account to bytes.
+    #[allow(clippy::wrong_self_convention)]
+    pub fn to_bytes(&self) -> [u8; Self::SIZE] {
+        match self {
+            Account::External(pk) => {
+                let mut bytes = [0u8; Self::SIZE];
+                let pk_bytes = pk.to_raw_bytes();
+
+                bytes[0] = 0;
+                bytes[1..].copy_from_slice(&pk_bytes);
+
+                bytes
+            }
+            Account::Contract(contract) => {
+                let mut bytes = [0u8; Self::SIZE];
+                let contract_bytes = contract.to_bytes();
+
+                bytes[0] = 1;
+                bytes[1..1 + contract_bytes.len()].copy_from_slice(&contract_bytes);
+
+                bytes
+            }
+        }
+    }
+}
+
+impl From<PublicKey> for Account {
+    fn from(pk: PublicKey) -> Self {
+        Self::External(pk)
+    }
+}
+
+impl From<ContractId> for Account {
+    fn from(contract: ContractId) -> Self {
+        Self::Contract(contract)
+    }
+}
+
+// The implementations of `PartialOrd` and `Ord`, while technically meaningless,
+// are extremely useful for using `Account` as keys of a `BTreeMap` in the
+// contract.
+
+impl PartialOrd for Account {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Account {
+    fn cmp(&self, other: &Self) -> Ordering {
+        use Account::*;
+
+        match (self, other) {
+            (External(lhs), External(rhs)) => {
+                let lhs = lhs.to_raw_bytes();
+                let rhs = rhs.to_raw_bytes();
+                lhs.cmp(&rhs)
+            }
+            (Contract(lhs), Contract(rhs)) => lhs.cmp(rhs),
+            // An externally owned account is defined as always "smaller" than a
+            // contract account. This ensures they are never mixed
+            // when ordering.
+            (External(_lhs), Contract(_rhs)) => Ordering::Greater,
+            (Contract(_lhs), External(_rhs)) => Ordering::Less,
+        }
+    }
+}
+
+/// The data an account has in the contract.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Archive, Serialize, Deserialize)]
+#[archive_attr(derive(CheckBytes))]
+pub struct AccountInfo {
+    /// The balance of the account.
+    pub balance: u64,
+    /// The current nonce of the account. Use the current value +1 to perform
+    /// an interaction with the account.
+    pub nonce: u64,
+    /// Status of the account.
+    ///
+    /// # Variants
+    /// 0: No Status
+    /// 1: Frozen
+    /// 2: Blocked
+    // TODO: We want to have this as a `Role` enum soon, but for serialization we use u64 temporarily.
+    pub status: u64,
+}
+
+impl AccountInfo {
+    const FROZEN: u64 = 1;
+    const BLOCKED: u64 = 2;
```

Comment:
Code doesn't compile with the constants being private and `NO_STATUS` is missing.
```suggestion
    /// Account is cleared to do all types of operations.
    pub const NO_STATUS: u64 = 0;
    /// Account is not allowed to do transfers but may still receive funds.
    pub const FROZEN: u64 = 1;
    /// Account is may not initiate transfers or receive funds.
    pub const BLOCKED: u64 = 2;
```